### PR TITLE
Add Protractor Strategy

### PR DESCRIPTION
**Pull request by joshrowley**
Adds the protractor strategy. I originally tried to have it run `webdriver-manager` as well, but it added a lot of complexity to the first implementation to try managing that since I have to wait for the selenium server to boot up before running the protractor spec, so instead I detect when webdriver-manager isn't running and tell the user to start webdriver-manager themselves.

Also started extracting out dependency objects that our strategies can use to check for and install dependencies. Made one for phantomJS and protractor.

Couple remaining things:

- add nodejs dependency checker/installer
- get `output` for the protractor configured correctly from the `.results.json` file to return the correct `results` to push to learn

### CODE REVIEW

**Comment by joshrowley** [comment]
I'm gonna punt on adding the node dependency until after I make the karma strategy.

File: lib/learn_test/dependencies/phantomjs.rb [hunk 1]
```diff
@@ -33,6 +31,12 @@ def phantom_installed_on_mac?
       def phantom_installed_on_linux?
         phantom_installed?
       end
+      def self.check_installation
+        new.check_installation
+      end
+
+      def check_installation
```
Comment (loganhasson) [inline lib/learn_test/dependencies/phantomjs.rb]: what's this guy for?

File: lib/learn_test/dependencies/phantomjs.rb [hunk 2]
```diff
@@ -33,6 +31,12 @@ def phantom_installed_on_mac?
       def phantom_installed_on_linux?
         phantom_installed?
       end
+      def self.check_installation
```
Comment (loganhasson) [inline lib/learn_test/dependencies/phantomjs.rb]: spacing

**Comment by loganhasson** [comment]
so, i'm thinking it's probably worth it to automatically run/kill the webdriver manager. i'm thinking something along the lines of:

1. check to see if running (can just do a `ps aux | grep webdriver manager | awk '{ print $2 }'` and seeing if there's a pid returned)
2. if not running, use `IO.popen` (i think) to start it and grab the pid
2a. if you needed to start it, run step 1 again in a loop until it's running
3. run tests
4. kill it

i think this'll be especially important on nitrous, because i'm worried users will just keep the server running

File: lib/learn_test/dependencies/phantomjs.rb (hunk 2 above)
Comment (joshrowley) [inline lib/learn_test/dependencies/phantomjs.rb]: that method isn't even used anymore so i just rmed it

File: lib/learn_test/dependencies/phantomjs.rb (hunk 1 above)
Comment (joshrowley) [inline lib/learn_test/dependencies/phantomjs.rb]: i think i was refactoring some logic out of that somewhere else and didn't clean it up, rming

**Comment by joshrowley** [comment]
@loganhasson cool yeah, will implement in a future PR

**Pull request merged**